File: VoidSchach960/src/main/java/figures/Queen.java
package figures;

import image.FigureImage;
import helper.*;
import board.*;
/**
 * @author stephan
 */
public class Queen extends Figure
{
	private Bishop bishop;
	private Rock   rock;

	public Queen(FigureImage figureImage, boolean isWhite, Position startPosition)
	{
		super( figureImage,isWhite,startPosition,(byte)5 );
		bishop = new Bishop( null,isWhite,startPosition );
		rock   = new Rock(   null,isWhite,startPosition );
	}

	public void figureMoved( Move move )
	{
		bishop.figureMoved( move );
		rock.figureMoved(   move );
		super.figureMoved(  move );
	}

	public void undoMove( Position oldPosition )
	{
		bishop.undoMove( oldPosition );
		rock.undoMove(   oldPosition );
		super.undoMove(  oldPosition );
	}
	
	public boolean isReachable(Position to,BasicChessGameInterface game)
	{
		return bishop.isReachable( to,game ) || rock.isReachable( to,game );
	}
	
	public void getReachableMoves( BasicChessGameInterface game,BasicMoveIterator result )
	{
		bishop.getReachableMoves(game,result);
		rock.getReachableMoves( game,result );
	}
	
	public boolean isSelectable( SimpleChessBoardInterface game ) 
	{
		return rock.isSelectable( game ) || bishop.isSelectable( game );
	}
	
	public int countReachableMoves( BasicChessGameInterface game )
	{
		return rock.countReachableMoves( game ) + bishop.countReachableMoves( game );
	}
	
	public boolean isQueen()
	{
		return true;
	}
	
	protected String getType()
	{
		return "Queen";
	}
}

File: module-engine-internal/src/main/kotlin/voidchess/figures/Queen.kt
package voidchess.figures

import voidchess.board.BasicChessBoard
import voidchess.board.ChessBoard
import voidchess.board.check.BoundLine
import voidchess.board.check.CheckLine
import voidchess.board.getKing
import voidchess.board.move.Direction
import voidchess.board.move.Move
import voidchess.board.move.Position


internal class Queen(isWhite: Boolean, startPosition: Position) : Figure(isWhite, startPosition, FigureType.QUEEN, true, true) {

    override fun isReachable(toPos: Position, game: BasicChessBoard): Boolean {
        val direction = position.getDirectionTo(toPos) ?: return false

        forEachReachablePos(game, direction) {
            if (it.equalsPosition(toPos)) return true
        }

        return false
    }

    private inline fun forEachReachablePos(game: BasicChessBoard, informOf: (Position) -> Unit) {
        forEachReachablePos(game, Direction.UP, informOf)
        forEachReachablePos(game, Direction.LEFT, informOf)
        forEachReachablePos(game, Direction.DOWN, informOf)
        forEachReachablePos(game, Direction.RIGHT, informOf)
        forEachReachablePos(game, Direction.UP_RIGHT, informOf)
        forEachReachablePos(game, Direction.UP_LEFT, informOf)
        forEachReachablePos(game, Direction.DOWN_RIGHT, informOf)
        forEachReachablePos(game, Direction.DOWN_LEFT, informOf)
    }

    private inline fun forEachReachableTakeableEndPos(game: BasicChessBoard, informOf: (Position) -> Unit) {
        forReachableTakeableEndPos(game, Direction.UP, informOf)
        forReachableTakeableEndPos(game, Direction.LEFT, informOf)
        forReachableTakeableEndPos(game, Direction.DOWN, informOf)
        forReachableTakeableEndPos(game, Direction.RIGHT, informOf)
        forReachableTakeableEndPos(game, Direction.UP_RIGHT, informOf)
        forReachableTakeableEndPos(game, Direction.UP_LEFT, informOf)
        forReachableTakeableEndPos(game, Direction.DOWN_RIGHT, informOf)
        forReachableTakeableEndPos(game, Direction.DOWN_LEFT, informOf)
    }

    override fun getReachableMoves(game: BasicChessBoard, result: MutableCollection<Move>) {
        forEachReachablePos(game) {
            result.add(Move[position, it])
        }
    }

    override fun getReachableTakingMoves(game: BasicChessBoard, result: MutableCollection<Move>) {
        forEachReachableTakeableEndPos(game) {
            result.add(Move[position, it])
        }
    }

    override fun getReachableCheckingMoves(game: ChessBoard, result: MutableCollection<Move>) {
        val opponentKingPos = game.getKing(!isWhite).position
        val currentPos = position
        forEachReachablePos(game) {
            if(isReachable(it, game)) {
                val figureTaken = game.move(this, it)
                if(isReachable(opponentKingPos, game)) {
                    result.add(Move[currentPos, it])
                }
                game.undoMove(this, currentPos, figureTaken)
            }
        }
    }

    override fun getPossibleMovesWhileUnboundAndCheck(game: ChessBoard, checkLine: CheckLine, result: MutableCollection<Move>) {
        // a queen can intersect a check at up to three positions. but the overhead to check this upper bound is probably to high
        checkLine.posProgression.forEachReachablePos {checkInterceptPos->
            addMoveIfReachable(checkInterceptPos, game, result)
        }
    }

    override fun getPossibleMovesWhileBoundAndNoCheck(game: ChessBoard, boundLine: BoundLine, result: MutableCollection<Move>) {
        boundLine.possibleMovesToAttacker.forEachReachablePos {posBetweenThisAndAttacker->
            result.add(Move[position, posBetweenThisAndAttacker])
        }
        boundLine.possibleMovesToKing.forEachReachablePos {posBetweenThisAndKing->
            result.add(Move[position, posBetweenThisAndKing])
        }
    }

    override fun isSelectable(game: ChessBoard): Boolean {
        forEachReachablePos(game) {
            if (!isBound(it, game)) return true
        }
        return false
    }

    override fun countReachableMoves(game: BasicChessBoard): Int {
        var reachableMovesCount = 0
        forEachReachablePos(game) {
            reachableMovesCount++
        }
        return reachableMovesCount
    }
}
